diff -up libhugetlbfs-2.16/Makefile.orig libhugetlbfs-2.16/Makefile
--- libhugetlbfs-2.16/Makefile.orig	2014-05-22 13:19:38.292337738 +0200
+++ libhugetlbfs-2.16/Makefile	2014-05-22 13:22:50.721884798 +0200
@@ -84,17 +84,25 @@ CFLAGS += -DNO_ELFLINK
 else
 ifeq ($(ARCH),s390x)
 CC64 = gcc -m64
-CC32 = gcc -m31
 ELF32 = elf_s390
 ELF64 = elf64_s390
 TMPLIB64 = lib64
 TMPLIB32 = lib
 CUSTOM_LDSCRIPTS = no
+ifneq ($(BUILDTYPE),NATIVEONLY)
+CC32 = gcc -m31
+endif
+else
+ifeq ($(ARCH),s390)
+CC32 = gcc -m31
+TMPLIB32 = lib
+CFLAGS += -DNO_ELFLINK
 else
 $(error "Unrecognized architecture ($(ARCH))")
 endif
 endif
 endif
+endif
 endif
 endif
 endif
